pkg/joinserver/redis/registry.go

```diff
@@ -214,19 +215,27 @@ func (r *DeviceRegistry) set(tx *redis.Tx, uid string, gets []string, f func(pb
 				return nil, errInvalidIdentifiers
 			}
 		} else {
-			if err := ttnpb.ProhibitFields(sets,
-				"ids.application_ids",
-				"ids.dev_eui",
-				"ids.device_id",
-				"ids.join_eui",
-				"provisioner_id",
-				"provisioning_data",
-			); err != nil {
-				return nil, errInvalidFieldmask.WithCause(err)
-			}
 			if err := cmd.ScanProto(updated); err != nil {
 				return nil, err
 			}
+			if ttnpb.HasAnyField(sets, "ids.application_ids") && !pb.ApplicationIdentifiers.Equal(updated.ApplicationIdentifiers) {
+				return nil, errReadOnlyField.WithAttributes("field", "ids.application_ids")
+			}
+			if ttnpb.HasAnyField(sets, "ids.device_id") && pb.DeviceID != updated.DeviceID {
+				return nil, errReadOnlyField.WithAttributes("field", "ids.device_id")
+			}
+			if ttnpb.HasAnyField(sets, "ids.join_eui") && (pb.JoinEUI == nil || !updated.JoinEUI.Equal(*pb.JoinEUI)) {
```

Comment:
Assuming you make the change above it's better to be consistent with AS registry and just check if these are equal without the `nil` check